### ISSUE TITLE
empty queries return 1 null item

### ISSUE THREAD

**Issue by jpmtrabbold** [benhutchins/dyngoose#482]
It's me again, sorry. When using any query (`all()` or `primaryKey.query`, for example), if there are no matching results, instead of returning an empty array with 0 items, it returns an array with 1 `null` item. So, I'm currently using the follow workaround:

```typescript
const keys = await FarmApiKey.primaryKey.query({ farmId: id });

// this is the workaround, because the count is 0 but keys, as an array, has 1 null item        
if (keys.count === 0)
   return []

return keys.map(k => ({
   // ** here goes mapping as usual, that would break without the workaround because k would be null
}))
```

**Comment by benhutchins** [comment]
Is there possibly more to this? There was a similar bug sometime last year, which I believe was resolved. There are specs that verify this in `primary-key.spec.ts`:

```typescript
it('should return an empty array when no results match', async () => {
      const res = await primaryKey.query({
        id: 420,
      })

      expect(res.length).to.eq(0)
      expect(res.count).to.eq(0)
      expect(res.records.length).to.eq(0)
    })
```

I tested a few things locally and couldn't reproduce a result that included a `null` in the array.

**Comment by jpmtrabbold** [comment]
Hmmm interesting - maybe one additional test would be:

```typescript 
expect(res[0]).to.eq(undefined)
expect(res.map(i => i)[0]).to.eq(undefined)
``` 

I know it looks silly - but this is a custom array object. Could it be that the lengths and counts are zero, but when actually accessing the index or trying to map it, it's returning a `null` instead of an `undefined`, therefore causing an iteration of 1?

Here's another use case that also returns the null result:

```typescript
        const farmEntities = await Farm.search().all()

        // when the count is zero, `all()` is returning one `null` result for some reason, 
        // so I have to intentionally return an empty array
        if (farmEntities.count === 0)
            return []

        // maps entity to the dto and returns
        return farmEntities.map(farm => ({
            id: farm.id,
            name: farm.name,
            address: farm.address
        }))
```

**Comment by benhutchins** [comment]
@jpmtrabbold I've done additional investigation and I haven't been able to reproduce this. So I'm stuck in finding a helpful way to resolve that.

I've gone ahead and added additional unit tests and created a new suite of tests specific to the output array class, but I can't get a situation to occur where the array appears to have represent an array with a null item `[null]`. All of these unit tests pass without any code changed, but I've decided to add a few changes entirely based on a few assumptions.

Knowing that `new Array(1)` in JavaScript creates an array of `[undefined]` and not `[null]`, I know that the somehow that means that the constructor is not being called with an array containing a `null`. So I've added several `!= null` condition checks throughout the output class. I cannot tell if these are helping the situation, as all the unit tests I've added pass without any of the changes. These checks don't seem like they're helping, because there is no possible situation that would cause `tableClass.fromDynamo` to result in a `null` output, so there should never be a situation where `new QueryOutput` is called without having an array of table instances.

I'm hoping you can test the change and tell me if you can discover anything additional. You can checkout the code changes in https://github.com/benhutchins/dyngoose/tree/fix/bh/482

**Comment by jpmtrabbold** [comment]
Sorry mate - lots of projects happening at the same time. I couldn't test that, but looking at the code, it looks like the guards you put in place will definitely prevent that issue from happening. I'm happy with that and I'll give it a go when you release it. Thank you very much!

**Comment by benhutchins** [comment]
I've merged the fix I created for this, it is included as part of Dyngoose 3.0.0
